File: 0384_Shuffle an Array/ShuffleanArray.java
class Solution {
    private int[] original;

    Random rand = new Random();

    private int randRange(int min, int max) {
        return rand.nextInt(max - min) + min;
    }

    private void swapAt(int[] array,int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    public Solution(int[] nums) {
        original = nums.clone();
    }

    public int[] reset() {
        int[] array = original.clone();
        return array;
    }

    public int[] shuffle() {
        int[] array = original.clone();
        for (int i = 0; i < array.length; i++) {
            swapAt(array,i, randRange(i, array.length));
        }
        return array;
    }
}

File: 0384_Shuffle an Array/ShuffleanArray.cs
public class Solution {

    private int[] original;
    private int[] shuffled;
    public Solution(int[] nums) {
        original = new int[nums.Length];
        shuffled = new int[nums.Length];
        
        Array.Copy(nums, original, nums.Length);
        Array.Copy(nums, shuffled, nums.Length);
    }
    
    /** Resets the array to its original configuration and return it. */
    public int[] Reset() {
        return original;
    }
    
    /** Returns a random shuffling of the array. */
    public int[] Shuffle() {
        var random = new Random();
        for(int i=shuffled.Length-1;i>=0;i--)
        {
            var index = random.Next(i+1);
            var tmp = shuffled[i];
            shuffled[i] = shuffled[index];
            shuffled[index] = tmp;
        }
        
        return shuffled;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int[] param_1 = obj.Reset();
 * int[] param_2 = obj.Shuffle();
 */